src/hooks/useUniversalRouter.ts

```diff
@@ -0,0 +1,76 @@
+import { BigNumber } from '@ethersproject/bignumber'
+import { TransactionResponse } from '@ethersproject/providers'
+import { Percent } from '@uniswap/sdk-core'
+import { SwapRouter, UNIVERSAL_ROUTER_ADDRESS } from '@uniswap/universal-router-sdk'
+import { FeeOptions } from '@uniswap/v3-sdk'
+import { useWeb3React } from '@web3-react/core'
+import { useCallback } from 'react'
+import { InterfaceTrade } from 'state/routing/types'
+import { calculateGasMargin } from 'utils/calculateGasMargin'
+import isZero from 'utils/isZero'
+import { swapErrorToUserReadableMessage } from 'utils/swapErrorToUserReadableMessage'
+
+import { PermitSignature } from './usePermitAllowance'
+
+interface SwapOptions {
+  slippageTolerance: Percent
+  deadline?: BigNumber
+  permit?: PermitSignature
+  feeOptions?: FeeOptions
+}
+
+export function useUniversalRouterSwapCallback(trade: InterfaceTrade | undefined, options: SwapOptions) {
+  const { account, chainId, provider } = useWeb3React()
+
+  return useCallback(async (): Promise<TransactionResponse> => {
+    try {
+      if (!account) throw new Error('missing account')
+      if (!chainId) throw new Error('missing chainId')
+      if (!provider) throw new Error('missing provider')
+      if (!trade) throw new Error('missing trade')
+
+      const { calldata: data, value } = SwapRouter.swapERC20CallParameters(trade, {
+        slippageTolerance: options.slippageTolerance,
+        deadlineOrPreviousBlockhash: options.deadline?.toString(),
+        inputTokenPermit: options.permit,
+        fee: options.feeOptions,
+      })
+      const tx =
+        value && !isZero(value)
+          ? {
+              from: account,
+              to: UNIVERSAL_ROUTER_ADDRESS(chainId),
+              data,
+              value,
+            }
+          : {
+              from: account,
+              to: UNIVERSAL_ROUTER_ADDRESS(chainId),
+              data,
+            }
+
+      let gasEstimate: BigNumber
+      try {
+        gasEstimate = await provider.estimateGas(tx)
+      } catch (gasError) {
+        await provider.call(tx) // this should throw the actual error
+        throw new Error('unexpected issue with gas estimation; please try again')
+      }
+      const gasLimit = calculateGasMargin(gasEstimate)
+      const response = await provider.getSigner().sendTransaction({ ...tx, gasLimit })
+      return response
```

Comment:
nit:
```suggestion
      return await provider.getSigner().sendTransaction({ ...tx, gasLimit })
```

Reply:
This is intentional because I find the `return await` construct confusing, and am tempted to remove the `await` (see https://github.com/Uniswap/widgets/pull/321#discussion_r1041172682 lol). In this case the `await` is important so that any rejection occurs within the try-block.